llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp

```diff
@@ -1255,8 +1256,17 @@ bool LoopIdiomRecognize::processLoopStoreOfLoopLoad(StoreInst *SI,
 
   Value *StorePtr = SI->getPointerOperand();
   const SCEVAddRecExpr *StoreEv = cast<SCEVAddRecExpr>(SE->getSCEV(StorePtr));
-  unsigned StoreSize = DL->getTypeStoreSize(SI->getValueOperand()->getType());
-
+  Type *StoreType = SI->getValueOperand()->getType();
+  unsigned StoreSize = DL->getTypeStoreSize(StoreType);
+  auto PreserveTags = PreserveCheriTags::Unknown;
+  if (DL->isFatPointer(StoreType->getScalarType())) {
+    // Capabilities and vectors of capabilities need to preserve tags
+    PreserveTags = PreserveCheriTags::Required;
```

Comment:
Eventually, we may want to drop ::Required since it's basically the same as unknown and just keep ::Unncessary as an optimizer hint that memcpy can be inlined in a non-tag-preserving way.